AK/PrintfImplementation.h

```diff
@@ -283,166 +285,227 @@ ALWAYS_INLINE int print_signed_number(PutChFunc putch, char*& bufptr, int number
+    }
+    ALWAYS_INLINE int format_X(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<u32>()(ap), true, state.alternate_form, state.left_pad, state.zero_pad, state.field_width);
+    }
+    ALWAYS_INLINE int format_n(PutChFunc, char*&, int nwritten, const State&, ArgumentListT& ap) const
+    {
+        *NextArgument<int*>()(ap) = nwritten;
+        return 0;
+    }
+    ALWAYS_INLINE int format_p(PutChFunc putch, char*& bufptr, int, const State&, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<u32>()(ap), false, true, false, true, 8);
+    }
+    ALWAYS_INLINE int format_P(PutChFunc putch, char*& bufptr, int, const State&, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<u32>()(ap), true, true, false, true, 8);
+    }
+    ALWAYS_INLINE int format_percent(PutChFunc putch, char*& bufptr, int, const State&, ArgumentListT&) const
+    {
+        putch(bufptr, '%');
+        return 1;
+    }
+    ALWAYS_INLINE int format_w(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<int>()(ap), false, state.alternate_form, false, true, 4);
+    }
+    ALWAYS_INLINE int format_b(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        return print_hex(putch, bufptr, NextArgument<int>()(ap), false, state.alternate_form, false, true, 2);
+    }
+    ALWAYS_INLINE int format_c(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT& ap) const
+    {
+        char c = NextArgument<int>()(ap);
+        return print_string(putch, bufptr, &c, 1, state.left_pad, state.field_width, state.dot);
+    }
+    ALWAYS_INLINE int format_unrecognized(char format_op, const char* fmt, PutChFunc, char*&, int, const State&, ArgumentListT&) const
+    {
+        dbg() << "printf_internal: Unimplemented format specifier " << format_op << " (fmt: " << fmt << ")";
+        return 0;
+    }
+};
+
+template<typename T, typename V>
+struct VaArgNextArgument {
+    ALWAYS_INLINE T operator()(V ap) const
+    {
+        return va_arg(ap, T);
+    }
+};
+
+#define PRINTF_IMPL_DELEGATE_TO_IMPL(c)                        \
+    case* #c:                                                  \
+        ret += impl.format_##c(putch, bufptr, ret, state, ap); \
+        break;
+
+template<typename PutChFunc, template<typename T, typename U, template<typename X, typename Y> typename V> typename Impl = PrintfImpl, typename ArgumentListT = va_list, template<typename T, typename V = decltype(declval<ArgumentListT&>())> typename NextArgument = VaArgNextArgument>
+ALWAYS_INLINE int printf_internal(PutChFunc putch, char* buffer, const char*& fmt, ArgumentListT ap)
```

Comment:
This could, too, be moved to the type as a method

Reply:
then the rest of the codebase would need to be changed too, `printf_internal` is _the_ interface for printf, no one should be able to override it, otherwise, `PrintfImpl` is useless.

Reply:
(repeated below in the 'typename U' bit)